package NYT.NRpc.NProto;

option java_package = "ru.yandex.yt.rpc";
option java_multiple_files = true;

import "yt/core/misc/guid.proto";
import "yt/core/misc/error.proto";

////////////////////////////////////////////////////////////////////////////////

message TRequestHeader
{
    optional NYT.NProto.TGuid request_id = 1;
    required string service = 2;
    required string method = 3;
    optional NYT.NProto.TGuid realm_id = 8;

    optional int32 protocol_version = 9 [default = 0];

    optional int64 start_time = 7;
    optional bool retry = 11 [default = false];
    optional NYT.NProto.TGuid mutation_id = 13;

    optional int64 timeout = 10;

    optional string user = 12;

    extensions 100 to max;
}

message TTracingExt
{
    extend NRpc.NProto.TRequestHeader
    {
        optional TTracingExt tracing_ext = 106;
    }

    optional uint64 trace_id = 1;
    optional uint64 span_id = 2;
    optional uint64 parent_span_id = 3;
}

message TBalancingExt
{
    extend NRpc.NProto.TRequestHeader
    {
        optional TBalancingExt balancing_ext = 111;
    }

    optional bool enable_stickness = 1 [default = false];
    optional int32 sticky_group_size = 2 [default = 1];
}

////////////////////////////////////////////////////////////////////////////////

message TResponseHeader
{
    optional NYT.NProto.TGuid request_id = 1;
    
    // If omitted then OK is assumed.
    optional NYT.NProto.TError error = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TRequestCancelationHeader
{
    required NYT.NProto.TGuid request_id = 1;
    required string service = 2;
    required string method = 3;
    optional NYT.NProto.TGuid realm_id = 4;
}

////////////////////////////////////////////////////////////////////////////////

message TReqDiscover
{
    extensions 100 to max;
}

message TRspDiscover
{
    required bool up = 1;
    repeated string suggested_addresses = 2;
}

////////////////////////////////////////////////////////////////////////////////

